Experimenting with Python 3

timj opened the pull request.

timj committed:
Naive run of futurize

timj committed:
Use os.walk instead of os.path.walk

os.path.walk is not present on Python 3.

timj committed:
Fix some byte vs str anomalies

timj committed:
Do not use [] as a default value

It persists between calls.

timj committed:
Fix md5 usage

This matches the style used in the rest of the source code and
is portable to Python 3. The file is opened in binary mode.

timj committed:
Use string pattern match

get_text_contents() is documented to return a string, not bytes.

timj committed:
Always decode the file contents

Even if no BOM is found it is still required to attempt
to decode the bytes. Do this with the default decoder.

timj committed:
Use string regex not bytes regex

get_text_contents is documented to return a string and not bytes.

This change adds a new attribute with the non-byte version of the
regex.

timj committed:
Use the unicode pattern match

timj committed:
Try to fix to_bytes and to_str methods

timj committed:
Joing with a string not bytes

timj committed:
Do not return return value from shutil.copytree

It does not return a proper exit status but returns a string
depending on how the copy happened. No explicit return status
is documented. Treat copytree like copy2 and return 0 if it
succeeds (otherwise it will raise an exception).

timj committed:
Fix __nonzero__ issue

futurize renamed __nonzero__ to __bool__ but did not fix the
__bool__ definitions that called __nonzero__. This patch just
deletes those superfluous methods.

timj commented:
@russel with these patches I can run `python3 bootstrap.py` to completion. Running the tests is still a disaster but that is mainly because of byte vs string inconsistencies or generated python code not being compatible with python 3.

624 tests fail (out of 1181?) and 155 return no result.

There are still some items that seem to think they don't work on python3 (they have the standard scons header).

russel commented:
@timj This is sounding exciting :-)  The obvious question: do all the tests pass with this when run on Python 2.7.x  I am not sure what the minimum x is but I am sure it will be 8 or 9. Since we do not have CI support for PRs here, I will take a clone of your repository and run a few checks.

russel commented:
@timj It only just struck me that we are here on GitHub using Git, the SCons mainline repository is a Mercurial one, should we be ensuring there is a path back to a mercurial repository to create the pull request for final merging into the mainline?

russel commented:
@timj  It just struck me that I don't use bootstrap.py. I cannot remember what the argument/discussion was but to execute SCons directly from a repository (which is my standard set up) I use an alias:

alias scons='/usr/bin/python /home/users/russel/Repositories/Mercurial/SCons/src/script/scons.py'

russel commented:
I tried running this branch on python 2.7.11 and got a whole mass of errors such as:

```
Traceback (most recent call last):
  File "test/Climb/option-u.py", line 36, in <module>
    test = TestSCons.TestSCons()
  File "/home/users/russel/Repositories/Git/SCons_TimJ/QMTest/TestSCons.py", line 269, in __init__
    import SCons.Node.FS
  File "/home/users/russel/Repositories/Git/SCons_TimJ/src/engine/SCons/Node/__init__.py", line 345, in <module>
    class NodeInfoBase(object):
TypeError: Error when calling the metaclass bases
    __weakref__ slot disallowed: either we already got one, or __itemsize__ != 0
```

timj commented:
Right. My main focus was working out exactly what is involved in python 3 support. I've been ignoring python2 in this experiment but of course we need it to work with both. Metaclasses are one of the things that changed quite a lot between versions.

dholth commented:
Thanks for doing this.

Too bad SCons is such a fan of opening files in binary mode. Luckly the % operator was added back to bytestrings. Actually the right thing to do would probably be to open in binary mode far less often.

Python Future has a great metaclass helper. http://python-future.org/what_else.html#metaclasses Where does SCons use them?

timj committed:
Do not add dict and weakref to slots when inheriting from object

future object seems to include __weakref__ and __dict__ in the
class hierarchy so can't add them in a subclass.

timj committed:
Also check basestring on python2

Temporary fix to python2 interaction with future str. The comments
in this file suggest basestring is slow but this is needed for now
until I work out how future str interacts with python2 str.

timj committed:
Use unicode literals

future str expects literals to be unicode strings on python2 but
"" is really bytes and is not a future str.

timj committed:
Fix tests to allow u"" to be present in repr output

williamblevins commented:
It seems that most of the python3 tests are failing do to str vs bytes issues. Can this be solved by not opening files in binary mode or does that cause other issues?

timj commented:
I have been pondering that. Ironically I think it's not helping that Scons has tried to fully support unicode in Python 2. Strings are the main porting headache. I am making some progress on getting this working with python 2 again but I'm having to enable unicode literals and a lot of the tests use `repr` and aren't expecting a `u` to turn up.

One problem I am finding is that python 2 literal strings (Without unicode literals enabled) do not play well with `future` `str` because that `str` is trying to use unicode semantics so `isinstance("abc", str)` returns false.

My current problem is that:

```
% python3 runtest.py src/engine/SCons/BuilderTests.py
```

passes all tests but:

```
% python2 runtest.py src/engine/SCons/BuilderTests.py
======================================================================
ERROR: test_emitter_TARGET_SOURCE (__main__.BuilderTestCase)
Test use of $TARGET and $SOURCE in emitter results
----------------------------------------------------------------------
Traceback (most recent call last):
  File "src/engine/SCons/BuilderTests.py", line 1343, in test_emitter_TARGET_SOURCE
    targets = builder(env, target = 'TTT', source ='SSS')
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Builder.py", line 633, in __call__
    return self._execute(env, target, source, OverrideWarner(kw), ekw)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Builder.py", line 554, in _execute
    tlist, slist = self._create_nodes(env, target, source)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Builder.py", line 532, in _create_nodes
    target=orig_tlist, source=orig_slist)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Environment.py", line 487, in arg2nodes
    v = node_factory(self.subst(v, **kw))
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Environment.py", line 518, in subst
    return SCons.Subst.scons_subst(string, self, raw, target, source, gvars, lvars, conv)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Subst.py", line 564, in scons_subst
    result = ss.substitute(strSubst, lvars)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Subst.py", line 528, in substitute
    result.append(self.conv(self.expand(a, lvars)))
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Util.py", line 459, in to_String_for_subst
    return str(s)
  File "/Users/timj/work/lsstsw/miniconda/lib/python2.7/site-packages/future/types/newstr.py", line 102, in __new__
    return super(newstr, cls).__new__(cls, value)
  File "/Users/timj/work/lsstsw/miniconda/lib/python2.7/site-packages/future/types/newobject.py", line 78, in __unicode__
    s = type(self).__str__(self)
  File "/Users/timj/work/lsst/scons/src/engine/SCons/Subst.py", line 258, in __str__
    return str(nl[0])
  File "/Users/timj/work/lsstsw/miniconda/lib/python2.7/site-packages/future/types/newstr.py", line 102, in __new__
    return super(newstr, cls).__new__(cls, value)
  File "/Users/timj/work/lsstsw/miniconda/lib/python2.7/site-packages/future/types/newobject.py", line 78, in __unicode__
    s = type(self).__str__(self)
TypeError: 'Delegate' object is not callable
```

williamblevins commented:
Is that with github up-to-date or newer commits on your end? About 680 python3 tests fail for me mostly due to str vs bytes issues in QMTest/TestCmd.py when calling the match functions though there are some other issues.

williamblevins commented:
I reread your message and it makes more sense now. I had the same issue with Future str in SCons.Util.is_string.

timj commented:
I have a couple of un-pushed patches. I'll push them in a bit.

timj committed:
Use default encoding and not specifically ascii

timj committed:
In py3 do not seem to be able to use locals() in list comprehension

timj committed:
Use safe_intern rather than sys.intern

timj committed:
test __bool__ not __nonzero__

timj committed:
Must encode trace strings before passing to write()

timj committed:
Use Util.is_String rather than local implementation

timj committed:
If somehow we have a string and not bytes, return it straightaway
